src/main/kotlin/fi/metatavu/muisti/devices/ExhibitionDeviceGroupController.kt

```diff
@@ -102,6 +121,45 @@ class ExhibitionDeviceGroupController {
     return result
   }
 
+  /**
+   * Copies device group including all it's contents (devices, content versions, group content versions and pages).
+   *
+   * @param sourceDeviceGroup source device group
+   * @param namePrefix prefix for copied resource names
+   * @param creatorId id of user that created the copy
+   * @return copied device group
+   */
+  fun copyDeviceGroup(
+    sourceDeviceGroup: ExhibitionDeviceGroup,
+    namePrefix: String,
+    creatorId: UUID
+  ): ExhibitionDeviceGroup {
+    logger.info("Creating copy of device group {}", sourceDeviceGroup.id)
+
+    val idMapper = IdMapper()
+
+    val id = idMapper.assignId(sourceDeviceGroup.id) ?: throw CopyException("Could not assign target source group id")
+    val exhibition = sourceDeviceGroup.exhibition ?: throw CopyException("Source device group exhibition not found")
+
+    val targetDeviceGroup = exhibitionDeviceGroupDAO.create(
+      id,
+      exhibition = exhibition,
+      room = sourceDeviceGroup.room ?: throw CopyException("Source device group room not found"),
+      name = "$namePrefix${sourceDeviceGroup.name}",
+      allowVisitorSessionCreation = sourceDeviceGroup.allowVisitorSessionCreation ?: throw CopyException("Source device group allowVisitorSessionCreation not found"),
+      visitorSessionEndTimeout = sourceDeviceGroup.visitorSessionEndTimeout ?: throw CopyException("Source device group visitorSessionEndTimeout not found"),
+      visitorSessionStartStrategy = sourceDeviceGroup.visitorSessionStartStrategy ?: throw CopyException("Source device group visitorSessionStartStrategy not found"),
+      creatorId = creatorId,
+      lastModifierId = creatorId
+    )
+
+    copyResources(exhibition, sourceDeviceGroup, targetDeviceGroup, namePrefix, idMapper, creatorId)
```

Comment:
Should we check that this function does not fail before returning?

Reply:
It throws a CopyException if it fails

Reply:
Ah, roger